src/Data/Constraint/Deriving/ClassDict.hs

```diff
@@ -208,6 +209,11 @@ mapResultType f t
   | (bndrs@(_:_), t') <- splitForAllTys t
     = mkSpecForAllTys bndrs $ mapResultType f t'
   | Just (at, rt) <- splitFunTy_maybe t
-    = mkFunTy at (mapResultType f rt)
+#if __GLASGOW_HASKELL__ >= 810
+    = mkVisFunTy
```

Comment:
Same problem as in `replaceTypeOccurrences`. Perhaps, it's worth a helper function?..